lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -56,14 +56,216 @@ removeNode = (node) ->
   removedNode = node.parentNode.removeChild(node)
   triggerEvent('page:after-node-removed', removedNode)
 
+updateHeadOrRefresh = (url, upstreamDocument, callback) ->
+  activeAssets = extractTrackedAssets(document)
+  upstreamAssets = extractTrackedAssets(upstreamDocument)
+  updateLinkTags(activeAssets, upstreamAssets)
+  updateScriptTagsOrRefresh(url, activeAssets, upstreamAssets, callback)
+
+extractTrackedAssets = (doc) ->
+  for node in doc.head.childNodes when node.getAttribute?('data-turbolinks-track')?
+    node
+
+insertScriptTask = (scriptNode) ->
+  newNode = document.createElement('SCRIPT')
+  newNode.setAttribute attr.name, attr.value for attr in scriptNode.attributes
+  newNode.setAttribute 'async', false
+  newNode.appendChild document.createTextNode scriptNode.innerHTML
+
+  return (done) ->
+    onScriptEvent = ->
+      newNode.removeEventListener('load', onScriptEvent)
+      newNode.removeEventListener('error', onScriptEvent)
+      done()
+    newNode.addEventListener('load', onScriptEvent)
+    newNode.addEventListener('error', onScriptEvent)
+    document.head.appendChild(newNode)
+    triggerEvent('page:after-script-inserted', newNode)
+
+filterForNodeType = (nodeType) ->
+  (node) -> node.nodeName == nodeType
+
+asyncSeries = (tasks, callback) ->
+  return callback() if tasks.length == 0
+  task = tasks.shift()
+  task(-> asyncSeries(tasks, callback))
+
+updateScriptTagsOrRefresh = (url, activeAssets, upstreamAssets, callback) ->
+  activeScriptTags = activeAssets.filter(filterForNodeType('SCRIPT'))
+  upstreamScriptTags = upstreamAssets.filter(filterForNodeType('SCRIPT'))
+  hasNewSrc = (upstreamNode) ->
+    !(activeScriptTags.some (activeNode) ->
+      upstreamNode.src == activeNode.src)
+  hasExistingScriptAssetName = (upstreamNode) ->
+    activeScriptTags.some (activeNode) ->
+      upstreamNode.attributes['data-turbolinks-track-script-as'].value == activeNode.attributes['data-turbolinks-track-script-as'].value
```

Comment:
Both scripts and links can use `data-turbolinks-track` but only script tags need to use `data-turbolinks-track-script-as` since we need a unique identifier across varied `src` attributes only for forcing refreshes on script version updates.

No script should ever use `data-turbolinks-track` without `data-turbolinks-track-script-as` for this reason.